src/terms/local.jl

```diff
@@ -71,19 +71,35 @@ function (external::ExternalFromFourier)(basis::PlaneWaveBasis{T}) where {T}
 end
 
 # Returns the form factors at unique values of |G + q| (in Cartesian coordinates).
-# Uses a hash map for O(1) lookup.
-function atomic_local_form_factors(basis::PlaneWaveBasis{T}, Gqs_cart::AbstractArray) where{T}
-    form_factors = IdDict{Tuple{Int,T},T}()  # IdDict for Dual compatibility
-    for G in Gqs_cart
+# Additionally, returns a mapping from any G index to a unique |G + q| index
+function atomic_local_form_factors(basis::PlaneWaveBasis{T}; q=zero(Vec3{T})) where{T}
+    Gqs_cart = [basis.model.recip_lattice * (G + q) for G in to_cpu(G_vectors(basis))]
+
+    iG2inorm_cpu = zeros(Int, length(Gqs_cart))
+    norm_indices = IdDict{T, Int}()
+    for (iG, G) in enumerate(Gqs_cart)
+        p = norm(G)
+        get!(norm_indices, p, length(norm_indices) + 1)
+        iG2inorm_cpu[iG] = norm_indices[p]
```

Comment:
```suggestion
        iG2inorm_cpu[iG] = get!(norm_indices, norm(G), length(norm_indices) + 1)
```
or is that too loaded of a line ?